src/controllers/rendering/controllerrenderingengine.cpp

```diff
@@ -0,0 +1,394 @@
+    m_waitCondition.wakeAll();
+}
+
+void ControllerRenderingEngine::finish() {
+    disconnect(this);
+    m_isValid = false;
+
+    if (m_context && m_context->isValid()) {
+        disconnect(m_context.get(),
+                &QOpenGLContext::aboutToBeDestroyed,
+                this,
+                &ControllerRenderingEngine::finish);
+        m_context->makeCurrent(m_offscreenSurface.get());
+        m_renderControl.reset();
+
+        std::shared_ptr<QOffscreenSurface> pOffscreenSurface = std::move(m_offscreenSurface);
+        QMetaObject::invokeMethod(
+                qApp,
+                [pOffscreenSurface] {
+                    pOffscreenSurface->destroy();
+                });
+        m_quickWindow.reset();
+
+        // Free the engine and FBO
+        m_fbo.reset();
+
+        m_context->doneCurrent();
+    }
+    m_context.reset();
+    m_pThread->quit();
+}
+
+void ControllerRenderingEngine::renderFrame() {
+    ScopedTimer t(u"ControllerRenderingEngine::renderFrame");
+    if (!m_isValid) {
+        DEBUG_ASSERT(!"Trying to render frame on an invalid engine");
+        return;
+    }
+
+    VERIFY_OR_TERMINATE(m_offscreenSurface->isValid(), "OffscreenSurface isn't valid anymore.");
+    VERIFY_OR_TERMINATE(m_context->isValid(), "GLContext isn't valid anymore.");
+    VERIFY_OR_TERMINATE(m_context->makeCurrent(m_offscreenSurface.get()),
+            "Couldn't make the GLContext current to the OffscreenSurface.");
+
+    if (!m_fbo) {
+        ScopedTimer t(u"ControllerRenderingEngine::renderFrame::initFBO");
+        VERIFY_OR_TERMINATE(
+                QOpenGLFramebufferObject::hasOpenGLFramebufferObjects(),
+                "OpenGL doesn't support FBO");
+
+        m_fbo = std::make_unique<QOpenGLFramebufferObject>(
+                m_screenInfo.size, QOpenGLFramebufferObject::CombinedDepthStencil);
+
+        GLenum glError;
+        glError = m_context->functions()->glGetError();
+
+        VERIFY_OR_TERMINATE(glError == GL_NO_ERROR, "GLError: " << glError);
+        VERIFY_OR_TERMINATE(m_fbo->isValid(), "Failed to initialize FBO");
+
+        m_quickWindow->setGraphicsDevice(QQuickGraphicsDevice::fromOpenGLContext(m_context.get()));
+
+        VERIFY_OR_TERMINATE(m_renderControl->initialize(),
+                "Failed to initialize redirected Qt Quick rendering");
+
+        m_quickWindow->setRenderTarget(QQuickRenderTarget::fromOpenGLTexture(m_fbo->texture(),
+                m_screenInfo.size));
+
+        m_quickWindow->setGeometry(0, 0, m_screenInfo.size.width(), m_screenInfo.size.height());
+    }
+
+    m_nextFrameStart = Clock::now();
+
+    m_renderControl->beginFrame();
+
+    if (m_pControllerEngine) {
+        m_pControllerEngine->pause();
+    }
+
+    m_renderControl->polishItems();
+
+    {
+        ScopedTimer t(u"ControllerRenderingEngine::renderFrame::sync");
+        VERIFY_OR_DEBUG_ASSERT(m_renderControl->sync()) {
+            kLogger.warning() << "Couldn't sync the render control.";
```

Comment:
It's been a long time now so I cannot recall the exact details, but I believe that a `false` in that case meant there was an issue with the sync. I remembered adding it because the scene would freeze in this sync would start returning false, before I implemented the lock mechanism. I'm happy to remove it, but I can tell that this is always returning `true`, even if there is no user induced scene graph update (you can test it with the dummy screen) 
Happy to remove the tear down procedure, but I would suggest we keep the assert/warning

Reply:
Updated with above